Add atstccfg meta config

rob05c opened the pull request:
## What does this PR (Pull Request) do?

Adds meta config to atstccfg config generator.

- [x] This PR is not related to any other Issue

Includes tests.
Includes changelog.
No documentation, no interface change.

## Which Traffic Control components are affected by this PR?
- Traffic Ops
- Traffic Ops ORT

## What is the best way to verify this PR?

Run unit tests.

Run `atstccfg` with request to the meta config endpoint, verify response is identical to Go and old Perl endpoints.

## If this is a bug fix, what versions of Traffic Control are affected?

Not a bug fix.

## The following criteria are ALL met by this PR
- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

alficles commented on lib/go-atscfg/meta.go:
```diff
@@ -0,0 +1,195 @@
+			} else { // ATSConfigMetaDataConfigFileScopeServers
+				scopeID = server.HostName
+			}
+			atsCfg.APIURI = "/api/1.2/" + string(scope) + "/" + scopeID + "/configfiles/ats/" + cfgFile
```
Hardcoded 1.2?

rob05c replied:
The config generator doesn't actually care what this number is. Unfortunately, the only place it's really defined is in the client, and we don't want this library to depend on that.
I changed it to a const.

alficles commented on lib/go-atscfg/meta.go:
```diff
@@ -0,0 +1,195 @@
+
+// getScope returns the ATSConfigMetaDataConfigFileScope for the given config file, and potentially the given server. If the config is not a Server scope, i.e. was part of an endpoint which does not include a server name or id, the server may be nil.
+func getScope(cfgFile string, scopeParams map[string]string) tc.ATSConfigMetaDataConfigFileScope {
+	switch {
```
Is the fallthrough here better than multiple-case? Like: `cfgFile == "ip_allow.config", cfgFile == "parent.config", ...`

rob05c replied:
Changed.

rawlinp commented on traffic_ops/traffic_ops_golang/ats/atsserver/meta.go:
```diff
@@ -85,136 +70,14 @@ func GetConfigMetaData(w http.ResponseWriter, r *http.Request) {
+	txt := atscfg.MakeMetaConfig(serverName, server, tmParams.URL, tmParams.ReverseProxyURL, locationParams, uriSignedDSes, scopeParams)
+	w.Header().Set(tc.ContentType, tc.ApplicationJson)
+	w.Write([]byte(txt))
+	api.WriteRespRaw(w, r, txt)
```
I know `atstccfg` isn't using this API, but I'm pretty sure this line should be removed since we're calling `w.Write` just above it.

rob05c replied:
Yeah, ORT/atstccfg doesn't use this anymore; but still a bug. Fixed.

rob05c commented:
>Also, I don't have a line to attach this comment to, but in routes.go I noticed the ATS meta config API route is duplicated (L436 and L469) if you'd like to remove one of those extra routes.

Fixed. Good catch! Guessing that was a non-conflict somewhere in all the merging around atstccfg PRs.

rob05c commented:
>I'm getting a TO API test failure on this:
Fixed. My fault for not running the API tests again after fixing Chris' version comment. And for making the API test brittle. I also fixed it to be less brittle.

rob05c commented:
Retest this please.
